### PR TITLE
Fix collection failures due to permission errors when using `--pyargs`

### LINKED ISSUE #11904: 8.0.0 + pyargs + permissions causes failures
Example of a successful test-install build on 7.4.4:

https://github.com/pulkin/pyteleport/actions/runs/7603885071/job/20706143053

On 8.0.0 it starts looking into inaccessible folders and fails right away:

https://github.com/pulkin/pyteleport/actions/runs/7733355863/job/21085241736

### PR DESCRIPTION
**Pull request by bluetech**
This should fix #11904. It does *not* fix all permission error situations that popped up with pytest 8.0.0 but most complaints were about `--pyargs` and others are mostly misconfigurations as far I could tell, so let's start with this to unblock pytest 8.1.0 and see about the others as the come.

The PR has some preparatory commits, here is the description of the main change (last commit):

In pytest<8, the collection tree for `pyargs` arguments in an invocation like this:

    pytest --collect-only --pyargs pyflakes.test.test_undefined_names

looked like this:

```xml
<Package test>
  <Module test_undefined_names.py>
    <UnitTestCase Test>
      <TestCaseFunction test_annotationUndefined>
      ... snipped ...
```

The pytest 8 collection improvements changed it to this:

```xml
<Dir pytest>
  <Dir .tox>
    <Dir venv>
      <Dir lib>
        <Dir python3.11>
          <Dir site-packages>
            <Package pyflakes>
              <Package test>
                <Module test_undefined_names.py>
                  <UnitTestCase Test>
                    <TestCaseFunction test_annotationUndefined>
                    ... snipped ...
```

Besides being egregious (and potentially even worse than the above, going all the way to the root, for system-installed packages, as is apparently common in CI), this also caused permission errors when trying to probe some of those intermediate directories.

This change makes `--pyargs` arguments no longer try to add parent directories to the collection tree according to the `--confcutdir` like they're regular arguments. Instead, only add the parents that are in the import path. This now looks like this:

```xml
<Package .tox/venv/lib/python3.11/site-packages/pyflakes>
  <Package test>
    <Module test_undefined_names.py>
      <UnitTestCase Test>
        <TestCaseFunction test_annotationUndefined>
        ... snipped ...
```

### CODE REVIEW

File: testing/test_collection.py [hunk 1]
```diff
@@ -1787,3 +1787,48 @@ def test_collect_short_file_windows(pytester: Pytester) -> None:
     test_file.write_text("def test(): pass", encoding="UTF-8")
     result = pytester.runpytest(short_path)
     assert result.parseoutcomes() == {"passed": 1}
+
+
+def test_pyargs_collection_tree(pytester: Pytester, monkeypatch: MonkeyPatch) -> None:
+    """When using `--pyargs`, the collection tree of a pyargs collection
+    argument should only parents in the import tree, not up to confcutdir.
```
Comment (nicoddemus) [inline testing/test_collection.py:1794]:
```suggestion
    argument should include only parents in the import tree, not up to confcutdir.
```

File: src/_pytest/main.py [hunk 2]
```diff
@@ -953,44 +968,64 @@ def genitems(
...
     if spec is None or spec.origin is None or spec.origin == "namespace":
-        return module_name
+        return None
     elif spec.submodule_search_locations:
         return os.path.dirname(spec.origin)
     else:
         return spec.origin
 
 
+@dataclasses.dataclass
```
Comment (nicoddemus) [inline src/_pytest/main.py:989]:
```suggestion
@dataclasses.dataclass(frozen=True)
```

**Review by nicoddemus — APPROVED** [review]
LGTM!

I specially liked the `CollectionArgument` dataclass, this is something that also caught my eye in the past.

Missing a CHANGELOG entry; I would add an example showing the old and the new collection tree, as it helps understand the change better.

**Comment by bluetech** [comment]
> Missing a CHANGELOG entry; I would add an example showing the old and the new collection tree, as it helps understand the change better.

I never forget to forget to `git add` the file, added now. I didn't add the collection tree comparisons, it seems like giving too much space for something not too significant. Instead, I linked to this PR for those interested in more info.

**Pull request merged**